src/main/java/teammates/sqllogic/api/SqlEmailGenerator.java

```diff
@@ -0,0 +1,1071 @@
+import teammates.storage.sqlentity.Student;
+
+/**
+ * Handles operations related to generating emails to be sent from provided templates.
+ *
+ * @see EmailTemplates
+ * @see EmailType
+ * @see EmailWrapper
+ */
+public final class SqlEmailGenerator {
+    // feedback action strings
+    private static final String FEEDBACK_ACTION_SUBMIT_EDIT_OR_VIEW = "submit, edit or view";
+    private static final String FEEDBACK_ACTION_VIEW = "view";
+    private static final String FEEDBACK_ACTION_SUBMIT_OR_UPDATE =
+                            ", in case you have not submitted yet or wish to update your submission. ";
+    private static final String HTML_NO_ACTION_REQUIRED = "<mark>No action is required if you have already submitted</mark>";
+
+    // status-related strings
+    private static final String FEEDBACK_STATUS_SESSION_OPEN = "is still open for submissions"
+                                            + FEEDBACK_ACTION_SUBMIT_OR_UPDATE + HTML_NO_ACTION_REQUIRED;
+    private static final String FEEDBACK_STATUS_SESSION_OPENING = "is now open";
+    private static final String FEEDBACK_STATUS_SESSION_CLOSING = "is closing soon"
+                                            + FEEDBACK_ACTION_SUBMIT_OR_UPDATE + HTML_NO_ACTION_REQUIRED;
+    private static final String FEEDBACK_STATUS_SESSION_CLOSED = "is now closed for submission";
+    private static final String FEEDBACK_STATUS_SESSION_OPENING_SOON = "is due to open soon";
+
+    private static final String DATETIME_DISPLAY_FORMAT = "EEE, dd MMM yyyy, hh:mm a z";
+
+    private static final long SESSION_LINK_RECOVERY_DURATION_IN_DAYS = 90;
+
+    private static final SqlEmailGenerator instance = new SqlEmailGenerator();
+
+    private final CoursesLogic coursesLogic = CoursesLogic.inst();
+    private final DeadlineExtensionsLogic deLogic = DeadlineExtensionsLogic.inst();
+    private final FeedbackSessionsLogic fsLogic = FeedbackSessionsLogic.inst();
+    private final UsersLogic usersLogic = UsersLogic.inst();
+
+    private SqlEmailGenerator() {
+        // prevent initialization
+    }
+
+    public static SqlEmailGenerator inst() {
+        return instance;
+    }
+
+    /**
+     * Generate Feedback Session Opening emails.
+     */
+    public List<EmailWrapper> generateFeedbackSessionOpeningEmails(FeedbackSession session) {
+        return generateFeedbackSessionOpeningOrClosingEmails(session, EmailType.FEEDBACK_OPENING);
+    }
+
+    private List<EmailWrapper> generateFeedbackSessionOpeningOrClosingEmails(
+            FeedbackSession session, EmailType emailType) {
+        Course course = session.getCourse();
+        boolean isEmailNeededForStudents = fsLogic.isFeedbackSessionForUserTypeToAnswer(session, false);
+        boolean isEmailNeededForInstructors = fsLogic.isFeedbackSessionForUserTypeToAnswer(session, true);
+        List<Instructor> instructorsToNotify = isEmailNeededForStudents
+                ? usersLogic.getCoOwnersForCourse(session.getCourse().getId())
```

Comment:
```suggestion
                ? usersLogic.getCoOwnersForCourse(course.getId())
```

Few more occurrences of this on the next few lines as well.